### DISCUSSION TITLE
Rust for STM32 Firmware?

### DISCUSSION THREAD

**Discussion by whoenig** [bitcraze/discussions#1192] [Q&A]
Is there any small example on how to write a function in Rust in the STM32 firmware and call this function from other existing C code?

Bonus question: can the written Rust code also be tested by including it in the Python bindings?

**Comment by ataffanel** [comment]
I made a minimal example 3 years ago, I just tested it and it still compiles and run with the latest Rust compiler :-): https://github.com/ataffanel/crazyflie-app-hello-rs.

This seems to have been done before the kconfig change so updating the firmware to the latest version might require some work. Though, the 3 years old firmware that is linked still works with a modern nRF firmware so it is workable for quick experiments. If you get to port it to newer firmware do not hesitate to do a PR ;-).

This example is the simplest possible, C functions are declared manually, there is no automatic binding for FreeRTOS or the Crazyflie API. The basic idea is to get rust to generate a ```.a``` static lib, and then to link this lib together with the Crazyflie firmware. This allows to make function callable by C (```pub extern "C" fn appMain() -> i32 { ... }```) and to call global C functions (```extern "C" { pub fn consolePutchar(ch: i32) -> i32; }```).

So, it should fit well with the rust binding as it is "only" a lib. Rust has a very good Python binding lib ([PyO3](https://pyo3.rs/)), but this would be to be used on the other side (linking the Crazyflie firmware into a Rust lib that exposes functionality to python), so this is not what we want here.
